core/templates/pages/admin-page/features-tab/admin-features-tab.component.spec.ts

```diff
@@ -0,0 +1,372 @@
+
+
+describe('Admin page feature tab', function() {
+  let component: AdminFeaturesTabComponent;
+  let fixture: ComponentFixture<AdminFeaturesTabComponent>;
+
+  let paramFactory: PlatformParameterObjectFactory;
+  let adminDataService: AdminDataService;
+  let featureApiService: PlatformFeatureAdminBackendApiService;
+  let adminTaskManagerService: AdminTaskManagerService;
+  let windowRef: WindowRef;
+
+  let updateApiSpy: jasmine.Spy;
+
+  let mockConfirmResult: (val: boolean) => void;
+  let mockPromptResult: (msg: string) => void;
+
+  beforeEach(async(() => {
+    TestBed
+      .configureTestingModule({
+        imports: [FormsModule, HttpClientTestingModule],
+        declarations: [AdminFeaturesTabComponent],
+      })
+      .compileComponents();
+
+    fixture = TestBed.createComponent(AdminFeaturesTabComponent);
+    component = fixture.componentInstance;
+
+    paramFactory = TestBed.get(PlatformParameterObjectFactory);
+    adminDataService = TestBed.get(AdminDataService);
+    featureApiService = TestBed.get(PlatformFeatureAdminBackendApiService);
+    windowRef = TestBed.get(WindowRef);
+    adminTaskManagerService = TestBed.get(AdminTaskManagerService);
+
+    let confirmResult = true;
+    let promptResult = 'mock msg';
+    spyOnProperty(windowRef, 'nativeWindow').and.returnValue({
+      confirm: () => confirmResult,
+      prompt: () => promptResult
+    });
+    mockConfirmResult = val => confirmResult = val;
+    mockPromptResult = msg => promptResult = msg;
+
+    spyOn(adminDataService, 'getDataAsync').and.resolveTo(<AdminPageData>{
+      featureFlags: [
+        paramFactory.createFromBackendDict({
+          data_type: 'bool',
+          default_value: false,
+          description: 'This is a dummy feature flag.',
+          feature_stage: FeatureStage.DEV,
+          is_feature: true,
+          name: 'dummy_feature',
+          rule_schema_version: 1,
+          rules: [{
+            filters: [
+              {
+                type: PlatformParameterFilterType.ServerMode,
+                conditions: [['=', ServerMode.Dev]]
+              }
+            ],
+            value_when_matched: true,
+          }],
+        })
+      ]
+    });
+
+    updateApiSpy = spyOn(featureApiService, 'updateFeatureFlag')
+      .and.resolveTo(null);
+
+    component.ngOnInit();
+  }));
+
+  it('should load feature flags on init', () => {
+    expect(component.featureFlags.length).toBe(1);
+    expect(component.featureFlags[0].name).toEqual('dummy_feature');
+  });
+
+  describe('.addNewRuleToTop', () => {
+    it('should add new rule to top of rule list', () => {
+      const featureFlag = component.featureFlags[0];
+
+      expect(featureFlag.rules.length).toBe(1);
+
+      component.addNewRuleToTop(featureFlag);
+      expect(featureFlag.rules.length).toBe(2);
+      expect(featureFlag.rules[0].valueWhenMatched).toBeFalse();
+    });
+  });
+
+  describe('.addNewRuleToBottom', () => {
```

Comment:
Per this, the above test, and other tests based on indexes or directionality: are these actually verifying that the correct thing is being added/removed/changed at the correct location? If not, please make sure they are. It seems like you could reverse the directionality of the prod functions for addNewRuleTo{Top,Bottom} and these tests would still pass.

Reply:
Improved these changes by adding more rules/filter/conditions, and checking that only the intended ones are changed and the others are untouched.